syncoid hold

rbike opened the pull request:
Added hold feature to syncoid.
Parameter "--use-hold" sets hold for newsyncsmap and removes hold from matchingsnap both on source and target. Hold name is "syncoid" + identitier + host name.
Setting holds on newsyncsnap prevents deletion of the of this snap until the next replication. In case of multiple targets a hold will be set for each target if different identifiers are used for each target. This prevents deletion until a new replications has been performed for all targets.

jimsalterjrs commented:
Does this look good to you, @phreaker0 ? I haven't operationally tested it yet, but looks good at first blush.

jimsalterjrs commented:
@rbike this will need updates to the in-program docs (which you see if you type `syncoid` with no arguments) as well, please.

rbike commented:
> @rbike this will need updates to the in-program docs (which you see if you type `syncoid` with no arguments) as well, please.

@jimsalterjrs  I updated the in-program docs and readme.md with the --use-hold argument description. I also added code that checks that matchingsnap exists before removing the hold. In case of an initial replication a matchingsnap doesn't exist.

phreaker0 commented on syncoid:
```diff
@@ -3,6 +3,9 @@
 # this software is licensed for use under the Free Software Foundation's GPL v3.0 license, as retrieved
 # from http://www.gnu.org/licenses/gpl-3.0.html on 2014-11-17.  A copy should also be available in this
 # project's Git repository at https://github.com/jimsalterjrs/sanoid/blob/master/LICENSE.
+# 
+# 2021-03-03: redbike
+# support for zfs holds added
```
please remove this here, you will be credited in the changelog after the next release which includes the patch

phreaker0 commented on syncoid:
```diff
@@ -827,7 +831,42 @@ sub syncdataset {
 		}
 	}
-
+# if "--use-hold" parameter is used set hold on newsync snapshot and remove hold on matching snapshot both on source and target
```
indentation needed here

phreaker0 commented on syncoid:
```diff
@@ -1325,7 +1364,7 @@ sub buildsynccmd {
 	my ($rhost,$fs,$newsyncsnap,$isroot,@snaps) = @_;
 
-	my $fsescaped = escapeshellparam($fs);
+	my $fsescaped = escapeshellparam($fs);	
```
remove the added whitespace

phreaker0 commented on syncoid:
```diff
@@ -1972,6 +2011,7 @@ Options:
   --no-sync-snap        Does not create new snapshot, only transfers existing
   --keep-sync-snap      Don't destroy created sync snapshots
   --create-bookmark     Creates a zfs bookmark for the newest snapshot on the source after replication succeeds (only works with --no-sync-snap)
+  --use-hold            Adds a hold to the newest snapshot on the source and target after replication succeeds and removes the hold after the next succesful replication. The hold name incldues the identifier if set. This allows for separate holds in case of multiple targets
```
please also document in the README.md

rbike replied:
@phreaker0 an updated readme.md is already provided

phreaker0 replied:
sry!, I somehow missed that

phreaker0 reviewed (changes requested):
I played around with it and it works fine, just some minor cosmetic stuff needs to be adapted as commented.

rbike commented:
Implemented all changes proposed by @phreaker0. Should be ready for merging.

rbike commented:
Hi @phreaker0 and @jimsalterjrs , is still some to be doner is it ready for merge?

phreaker0 reviewed (approved):
@rbike i rebased your PR to the latest master, thanks for the adaptions. Looks ready to me
